[Transforms] Refactor CreateControlFlowHub

ssahasra opened the pull request:
CreateControlFlowHub is a method that redirects control flow edges from a set of incoming blocks to a set of outgoing blocks through a new set of "guard" blocks. This is now refactored into a separate file with one enhancement: The input to the method is now a set of branches rather than two sets of blocks.

The original implementation reroutes every edge from incoming blocks to outgoing blocks. But it is possible that for some incoming block InBB, some successor S might be in the set of outgoing blocks, but that particular edge should not be rerouted. The new implementation makes this possible by allowing the user to specify the targets of each branch that need to be rerouted.

This is needed when improving the implementation of FixIrreducible #101386. Current uses in FixIrreducible and UnifyLoopExits do not demonstrate this finer control over the edges being rerouted.

ssahasra committed:
[Transforms] Refactor CreateControlFlowHub

CreateControlFlowHub is a method that redirects control flow edges from a set of
incoming blocks to a set of outgoing blocks through a new set of "guard" blocks.
This is now refactored into a separate file with one enhancement: The input to
the method is now a set of branches rather than two sets of blocks.

The original implementation reroutes every edge from incoming blocks to outgoing
blocks. But it is possible that for some incoming block InBB, some successor S
might be in the set of outgoing blocks, but that particular edge should not be
rerouted. The new implementation makes this possible by allowing the user to
specify the targets of each branch that need to be rerouted.

This is needed when improving the implementation of FixIrreducible #101386.
Current use in FixIrreducible does not demonstrate this finer control over the
edges being rerouted. But in UnifyLoopExits, when only one successor of an
exiting block is an exit block, this refinement now reroutes only the relevant
control-flow through the edge; the non-exit successor is not rerouted. This
results in fewer branches and PHI nodes in the hub.

arsenm commented on llvm/lib/Transforms/Utils/ControlFlowUtils.cpp:
```diff
@@ -0,0 +1,341 @@
+      auto Succ1Iter = find(Outgoing, Succ1);
+      Value *Id0 = ConstantInt::get(Type::getInt32Ty(Context),
+                                    std::distance(Outgoing.begin(), Succ0Iter));
+      Value *Id1 = ConstantInt::get(Type::getInt32Ty(Context),
```
Avoid repeated getInt32Ty calls

ssahasra replied:
Done.

arsenm commented on llvm/lib/Transforms/Utils/ControlFlowUtils.cpp:
```diff
@@ -0,0 +1,341 @@
+    Value *NewV = NewPhi;
+    if (AllUndef) {
+      NewPhi->eraseFromParent();
+      NewV = UndefValue::get(Phi->getType());
```
Use poison

ssahasra replied:
Done.

arsenm commented on llvm/lib/Transforms/Utils/ControlFlowUtils.cpp:
```diff
@@ -0,0 +1,341 @@
+                        Phi->getName() + ".moved", FirstGuardBlock->begin());
+    bool AllUndef = true;
+    for (auto [BB, Succ0, Succ1] : Incoming) {
+      Value *V = UndefValue::get(Phi->getType());
```
Poison

arsenm commented on llvm/lib/Transforms/Utils/ControlFlowUtils.cpp:
```diff
@@ -0,0 +1,341 @@
+    DTU->applyUpdates(Updates);
+  }
+
+  for (auto I : DeletionCandidates) {
+    if (I->use_empty())
+      if (auto *Inst = dyn_cast_or_null<Instruction>(I))
+        Inst->eraseFromParent();
+  }
```
I guess this isnt' a change, but does this really need to bother cleaning this up?

ssahasra replied:
It's nice to cleanup while we can, and results in cleaner lit tests since the unused PHIs show up as removed in the checks.

arsenm commented on llvm/lib/Transforms/Utils/UnifyLoopExits.cpp:
```diff
@@ -140,53 +141,43 @@ static void restoreSSA(const DominatorTree &DT, const Loop *L,
 
+static bool isExitBlock(Loop *L, BasicBlock *Succ, LoopInfo &LI) {
+  Loop *SL = LI.getLoopFor(Succ);
+  if (SL == L || L->contains(SL))
+    return false;
+  return true;
```
return bool expression

ssahasra replied:
Done.

arsenm commented on llvm/test/Transforms/FixIrreducible/switch.ll:
```diff
@@ -1,5 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
-; RUN: opt < %s -passes='lower-switch,fix-irreducible' -S | FileCheck %s
+; RUN: opt < %s -lowerswitch -fix-irreducible --verify-loop-info -S | FileCheck %s
```
Does this really need the old PM run lines?

ssahasra replied:
It doesn't really matter, but I have been copy-pasting existing lines, and kept this to be consistent with all the other tests.

arsenm commented on llvm/test/Transforms/FixIrreducible/switch.ll:
```diff
@@ -1,5 +1,6 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
-; RUN: opt < %s -passes='lower-switch,fix-irreducible' -S | FileCheck %s
+; RUN: opt < %s -lowerswitch -fix-irreducible --verify-loop-info -S | FileCheck %s
+; RUN: opt < %s -passes="lower-switch,fix-irreducible,verify<loops>" -S | FileCheck %s
```
I think these usually use single quotes

ssahasra replied:
Fixed in all tests.

ssahasra committed:
review comments:

- create Int32Ty once
- use poison instead of undef
- single quotes in lit tests
- the check for loop exit is actually very trivial

ssahasra commented:
I just pushed the squashed patch to main, I was not aware that pushing directly to main does not close the PR! Not sure how to mark this PR as submitted.
